File: codes/Garnetwzy/145.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ret;
        if(root == NULL)
            return ret;
        stack<TreeNode*> s;
        TreeNode *p = root;
        do{
            while(p) {
                s.push(p);
                p = p->left;
            }
            TreeNode *pre = NULL;
            bool check = true;
            while(!s.empty() && check) {
                if(s.top()->right == pre) {
                    ret.push_back(s.top()->val);
                    pre = s.top();
                    s.pop();
                }else {
                    p = s.top()->right;
                    check = false;
                }
            }
        }while(!s.empty());
        return ret;
    }
};

File: codes/HellDream/145.java
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        if(root==null) return list;
        Stack<TreeNode> stack = new Stack<>();
        TreeNode head = root;
        stack.push(root);
        while(!stack.isEmpty()){
            TreeNode node = stack.peek();
            if((node.left==null && node.right==null) ||node.left==head||node.right==head){
                list.add(node.val);
                stack.pop();
                head = node;
            }else{
                if(node.right!=null) stack.push(node.right);
                if(node.left!=null) stack.push(node.left);
            }
        }
        return list;
    }
}
     